File: src/init/scaffold.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

import { readBundledSkill } from "../bundled-skill.js";
import {
  CONFIG_FILENAME,
  configFileValidator,
  GEOMEAN_PRIMARY,
  loopKeyProblems,
} from "../config.js";
import { GymratError } from "../errors.js";
import { DEFAULT_RUNBOOK_PATH, type WizardResult } from "./wizard.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** The outcome of writing one scaffold artifact: newly written, already present, or user-declined. */
export type ArtifactStatus = "created" | "exists" | "declined";

/** The return shape of {@link scaffold}: one entry per artifact it may write. */
export interface ScaffoldResult {
  config: { path: string; status: "created" };
  runbook: { path: string; status: ArtifactStatus };
  skill: { path: string; status: ArtifactStatus };
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Path, relative to the project root, where init writes and doctor checks for the skill file. */
export const SKILL_RELATIVE_PATH = ".claude/skills/gymrat/SKILL.md";

const RUNBOOK_STUB = `# Optimization Runbook

## Goal

<!-- Describe the optimization goal here. -->

## Gating metrics

<!-- List the metrics that must not regress. -->

## Constraints

<!-- List any constraints on the optimization. -->

## Approaches to try

<!-- List strategies for the agent to explore. -->

\`gymrat supervise\` injects this file into the agent's instructions.
`;

// ---------------------------------------------------------------------------
// Config construction
// ---------------------------------------------------------------------------

function buildConfig(wizardResult: WizardResult): Record<string, unknown> {
  const config: Record<string, unknown> = { bench: wizardResult.bench };

  if (wizardResult.adapter !== undefined) {
    config.adapter = wizardResult.adapter;
  }
  if (wizardResult.checks !== undefined) {
    config.checks = wizardResult.checks;
  }
  if (wizardResult.primary !== undefined) {
    config.primary = wizardResult.primary;
  }

  const stop: Record<string, unknown> = {};
  if (wizardResult.stopTarget !== undefined) {
    stop.targetValue = wizardResult.stopTarget;
  }
  if (wizardResult.stopMaxIterations !== undefined) {
    stop.maxIterations = wizardResult.stopMaxIterations;
  }
  if (Object.keys(stop).length > 0) {
    config.stop = stop;
  }

  if (wizardResult.runbook !== false) {
    config.runbook = wizardResult.runbook.path;
  }

  return config;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

function validateConfig(config: Record<string, unknown>): void {
  if (!configFileValidator.check(config)) {
    const issue = configFileValidator.firstIssue(config);
    if (issue === undefined) throw new GymratError("validation failed but no issue was reported");
    throw new GymratError(`Invalid config: ${issue.path} must be ${issue.expected}`);
  }

  const problems = loopKeyProblems({
    primary: config.primary ?? GEOMEAN_PRIMARY,
    ...(config.stop !== undefined && { stop: config.stop }),
  });
  const firstProblem = problems[0];
  if (firstProblem !== undefined) {
    throw new GymratError(firstProblem);
  }
}

// ---------------------------------------------------------------------------
// File writers
// ---------------------------------------------------------------------------

function writeRunbook(
  baseDir: string,
  wizardResult: WizardResult,
): { path: string; status: ArtifactStatus } {
  if (wizardResult.runbook === false) {
    return { path: DEFAULT_RUNBOOK_PATH, status: "declined" };
  }

  const runbookPath = wizardResult.runbook.path;
  const fullPath = resolve(baseDir, runbookPath);

  if (existsSync(fullPath)) {
    return { path: runbookPath, status: "exists" };
  }

  mkdirSync(dirname(fullPath), { recursive: true });
  writeFileSync(fullPath, RUNBOOK_STUB);
  return { path: runbookPath, status: "created" };
}

function writeSkill(
  baseDir: string,
  wizardResult: WizardResult,
  importMetaUrl: string,
): { path: string; status: ArtifactStatus } {
  if (!wizardResult.installSkill) {
    return { path: SKILL_RELATIVE_PATH, status: "declined" };
  }

  const fullPath = join(baseDir, SKILL_RELATIVE_PATH);

  if (existsSync(fullPath)) {
    return { path: SKILL_RELATIVE_PATH, status: "exists" };
  }

  const content = readBundledSkill(importMetaUrl);
  mkdirSync(dirname(fullPath), { recursive: true });
  writeFileSync(fullPath, content);
  return { path: SKILL_RELATIVE_PATH, status: "created" };
}

function writeConfig(
  baseDir: string,
  config: Record<string, unknown>,
): { path: string; status: "created" } {
  const fullPath = join(baseDir, CONFIG_FILENAME);
  writeFileSync(fullPath, JSON.stringify(config, null, 2) + "\n");
  return { path: CONFIG_FILENAME, status: "created" };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Write the gymrat config, runbook stub, and skill file to `baseDir`.
 *
 * `importMetaUrl` is the caller's `import.meta.url`, forwarded to
 * `readBundledSkill` so the bundled skill can be resolved from the
 * package's install location.
 *
 * Write order is runbook -> skill -> config so a failure partway through
 * never leaves a `gymrat.json` pointing at artifacts that were not created.
 */
export function scaffold(
  baseDir: string,
  wizardResult: WizardResult,
  importMetaUrl: string,
): ScaffoldResult {
  const config = buildConfig(wizardResult);
  validateConfig(config);

  const runbook = writeRunbook(baseDir, wizardResult);
  const skill = writeSkill(baseDir, wizardResult, importMetaUrl);
  const configResult = writeConfig(baseDir, config);

  return { config: configResult, runbook, skill };
}

File: src/gymrat/init/scaffold.py
"""Write the gymrat config, runbook stub, and skill file for ``init``.

A ``ScaffoldRequest`` carries the three user choices (bench command, runbook
flag, skill-install flag). The config is serialized as hand-written TOML —
``json.dumps`` escapes the bench string, which round-trips through any TOML
parser because every JSON basic-string escape is valid TOML. Validation and
the bundled-skill read run before any file is written, so a broken install
leaves nothing behind.

The scaffold is re-runnable: an existing ``gymrat.toml`` is left byte-identical
and reported as ``exists``, while the runbook and skill are still filled in.
``bench`` is therefore only required when the config has to be written. If a
later artifact write fails, a config this run created is removed so no partial
scaffold is left — one that was already there is never touched.
"""

import json
import os
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from gymrat.bundled_skill import read_bundled_skill
from gymrat.config import CONFIG_FILENAME, validate_config_dict
from gymrat.errors import GymratError

#: Path, relative to the project root, where init writes and doctor checks for the skill.
SKILL_RELATIVE_PATH = ".claude/skills/gymrat/SKILL.md"

DEFAULT_RUNBOOK_PATH = "gymrat-runbook.md"

type ArtifactStatus = Literal["created", "exists", "declined", "is a directory"]

_RUNBOOK_STUB = """# Optimization Runbook

## Goal

<!-- Describe the optimization goal here. -->

## Gating metrics

<!-- List the metrics that must not regress. -->

## Constraints

<!-- List any constraints on the optimization. -->

## Approaches to try

<!-- List strategies for the agent to explore. -->

`gymrat supervise` injects this file into the agent's instructions.
"""


@dataclass(frozen=True, slots=True)
class ScaffoldRequest:
    """User choices that drive the scaffold: bench command, runbook, and skill."""

    bench: str | None = None
    runbook: bool = True
    install_skill: bool = True


@dataclass(frozen=True, slots=True)
class ScaffoldArtifact:
    """The outcome of one scaffold artifact: its path and whether it was written."""

    path: str
    status: ArtifactStatus


@dataclass(frozen=True, slots=True)
class ScaffoldResult:
    """One entry per artifact :func:`scaffold` may write."""

    config: ScaffoldArtifact
    runbook: ScaffoldArtifact
    skill: ScaffoldArtifact


def _serialize_config(config: dict[str, object]) -> str:
    """Hand-written TOML: one key per line, trailing newline.

    ``json.dumps`` escapes the bench value — every JSON basic-string escape is a
    valid TOML basic-string escape, so the result round-trips through any TOML
    parser.
    """
    lines = [f"bench = {json.dumps(config['bench'])}"]
    if "runbook" in config:
        lines.append(f'runbook = "{config["runbook"]}"')
    return "\n".join(lines) + "\n"


def _write_runbook(base_dir: str, *, runbook: bool) -> ScaffoldArtifact:
    if not runbook:
        return ScaffoldArtifact(path=DEFAULT_RUNBOOK_PATH, status="declined")

    full_path = Path(base_dir) / DEFAULT_RUNBOOK_PATH
    if full_path.exists():
        if not full_path.is_file():
            return ScaffoldArtifact(path=DEFAULT_RUNBOOK_PATH, status="is a directory")
        return ScaffoldArtifact(path=DEFAULT_RUNBOOK_PATH, status="exists")

    try:
        full_path.write_text(_RUNBOOK_STUB, encoding="utf-8")
    except OSError as exc:
        msg = f"Cannot write {DEFAULT_RUNBOOK_PATH} in {base_dir}"
        raise GymratError(msg, hint=str(exc)) from exc
    return ScaffoldArtifact(path=DEFAULT_RUNBOOK_PATH, status="created")


def _write_skill(base_dir: str, content: str) -> ScaffoldArtifact:
    full_path = Path(base_dir) / SKILL_RELATIVE_PATH
    if full_path.exists():
        if not full_path.is_file():
            return ScaffoldArtifact(path=SKILL_RELATIVE_PATH, status="is a directory")
        return ScaffoldArtifact(path=SKILL_RELATIVE_PATH, status="exists")

    try:
        full_path.parent.mkdir(parents=True, exist_ok=True)
        full_path.write_text(content, encoding="utf-8")
    except OSError as exc:
        msg = f"Cannot write {SKILL_RELATIVE_PATH} in {base_dir}"
        raise GymratError(msg, hint=str(exc)) from exc
    return ScaffoldArtifact(path=SKILL_RELATIVE_PATH, status="created")


def _prepare_config(request: ScaffoldRequest) -> str:
    """Validate the requested config and render it as TOML."""
    config_dict: dict[str, object] = {"bench": request.bench}
    if request.runbook:
        config_dict["runbook"] = DEFAULT_RUNBOOK_PATH
    validate_config_dict(config_dict)
    return _serialize_config(config_dict)


def _write_config(base_dir: str, content: str) -> ScaffoldArtifact:
    """Write the config via a temp file + ``os.replace`` for atomicity.

    A crash or permission error mid-write never leaves a truncated
    ``gymrat.toml`` — the temp file is cleaned up on failure.
    """
    full_path = Path(base_dir) / CONFIG_FILENAME
    fd = -1
    tmp_path: str | None = None
    try:
        fd, tmp_path = tempfile.mkstemp(prefix="gymrat.toml.", dir=base_dir, suffix=".tmp")
        os.write(fd, content.encode("utf-8"))
        os.close(fd)
        fd = -1
        Path(tmp_path).replace(full_path)
        tmp_path = None
    except OSError as exc:
        msg = f"Cannot write {CONFIG_FILENAME} in {base_dir}"
        raise GymratError(msg, hint=str(exc)) from exc
    finally:
        if fd >= 0:
            os.close(fd)
        if tmp_path is not None:
            Path(tmp_path).unlink(missing_ok=True)
    return ScaffoldArtifact(path=CONFIG_FILENAME, status="created")


def _path_blocked(base_dir: str, relative: str) -> bool:
    """True when a non-regular file occupies ``relative``.

    Symlinks (including dangling ones) are always blocked — writing through a
    symlink would place the content at a location the user did not choose.
    Directories are blocked because they cannot be opened as regular files.
    """
    full = Path(base_dir) / relative
    if full.is_symlink():
        return True
    return full.exists() and not full.is_file()


def _blocked_paths(base_dir: str, request: ScaffoldRequest) -> list[str]:
    """Collect artifact paths that are blocked by a non-regular file."""
    blocked: list[str] = []
    if _path_blocked(base_dir, CONFIG_FILENAME):
        blocked.append(CONFIG_FILENAME)
    if request.runbook and _path_blocked(base_dir, DEFAULT_RUNBOOK_PATH):
        blocked.append(DEFAULT_RUNBOOK_PATH)
    if request.install_skill and _path_blocked(base_dir, SKILL_RELATIVE_PATH):
        blocked.append(SKILL_RELATIVE_PATH)
    return blocked


def scaffold(base_dir: str, request: ScaffoldRequest) -> ScaffoldResult:
    """Write the config, runbook stub, and skill file for ``base_dir``.

    Returns a :class:`ScaffoldResult` describing each artifact. An existing
    ``gymrat.toml`` is reported as ``exists`` and left byte-identical; the
    remaining artifacts are still created, which makes a re-run the way to
    restore a deleted runbook or skill.

    Raises :class:`GymratError` when:

    - The config fails validation or the bundled skill cannot be read.
    - A non-regular file (directory, symlink) occupies an artifact path.
    - A filesystem error (permission denied, read-only FS) prevents writing.

    In every failure case no partial scaffold is left behind.
    """
    config_path = Path(base_dir) / CONFIG_FILENAME
    try:
        config_exists = config_path.exists()
    except OSError as error:
        msg = f"Cannot access {config_path}: {error}"
        raise GymratError(msg, hint="Check directory permissions.") from error

    config_content = None if config_exists else _prepare_config(request)
    skill_content = read_bundled_skill() if request.install_skill else None

    blocked = _blocked_paths(base_dir, request)
    if blocked:
        paths = ", ".join(blocked)
        msg = f"Blocked path: {paths}"
        raise GymratError(msg, hint="Remove or rename the blocking entry and re-run.")

    config_artifact = (
        ScaffoldArtifact(path=CONFIG_FILENAME, status="exists")
        if config_content is None
        else _write_config(base_dir, config_content)
    )
    try:
        runbook_artifact = _write_runbook(base_dir, runbook=request.runbook)
        skill_artifact = (
            ScaffoldArtifact(path=SKILL_RELATIVE_PATH, status="declined")
            if skill_content is None
            else _write_skill(base_dir, skill_content)
        )
    except BaseException:
        # Only roll back a config this run created; a pre-existing one is the user's.
        if config_content is not None:
            config_path.unlink(missing_ok=True)
        raise

    return ScaffoldResult(config=config_artifact, runbook=runbook_artifact, skill=skill_artifact)
